File: backend/app/utils/tournament_event_link.py
"""Which ``events`` row is this registered matchup — answered by id, never by name.

═══ WHY THIS MODULE EXISTS ═══

Alex, 2026-08-28: *"I thought that tournaments were containers for related
events."*  That is the ruled data model, and it is also what the database
holds.  UX-P149 built a bespoke match page on the premise that a tennis
matchup has nowhere to live — *"tennis matches have no `events` row (zero
exist for any registered matchup)"*.  That was true when it was measured and
is not true now: the Odds API began carrying US Open main-draw singles on
**2026-08-27 21:05 UTC**, and 94 standard ``events`` rows exist for the 96
registered R128 fixtures.  A whole parallel surface was built in the window
between the census and the ingest.

So a match is an event like any other, the tournament is curation on top, and
the click-through goes to ``/events/{id}``.  This module is the join that makes
that possible.

═══ THE CHANNEL IS AN ID, AND ONLY AN ID ═══

The obvious join — two player names and a date — is the one thing this module
must not do.  Gotcha #32 / ruling 048: an id-less claim never absorbs, because
a duplicate is visible and reversible and a wrong correspondence is neither.
A link is a weaker claim than an absorption, but it fails the same way: a
reader who taps "Djokovic vs Navone" and lands on a different match has been
lied to by a surface whose whole posture is that identity is pinned.

The id path already exists and needs nothing built:

    register matchup
      -> its pinned match-winner ``market_id``      (register-owned, or filled
                                                     by Q426's link overlay)
      -> ``futures_markets.event_id``               (set by the matching layer)
      -> ``events.id``

Every hop is a primary-key or indexed-column lookup.  No name comparison, no
time window, no category test — the same posture ``_load_match_group`` takes
for the sibling props, and for the same reason.

Measured against production 2026-08-28: 134 Kalshi/Polymarket markets carry an
``event_id`` pointing at one of the 94 US Open events, covering **90** of them.
The six R128 fixtures that do not dereference get no link, which is the correct
output: a missing link is a smaller harm than a wrong one, and
``unresolved_reasons`` counts them by name so the gap is never silent.

═══ WHAT IT REFUSES ═══

* ``NO_PINNED_MARKET``  — the matchup's source blocks pin no market id (the
  committed register's R128 blocks are all ``status: "missing"`` until the
  Q426 overlay fills them).
* ``MARKET_NOT_FOUND``  — a pinned id with no row.  Pinned identity that has
  gone stale is a register finding, not something to route around.
* ``MARKET_UNLINKED``   — the market exists and carries no ``event_id``.  The
  matching layer has not claimed it; this module does not get to guess.
* ``EVENT_DISAGREEMENT`` — two pinned markets for one matchup dereference to
  two different events.  Exactly the case where a name-matcher would pick one
  and look right.  Both are dropped: if the ids disagree, we do not know.

═══ THE SECOND CHANNEL: THE AUTHORITY'S OWN EVENT ID (#2693 step 2) ═══

The channel above starts at a *register matchup*, and ``build_slate`` retires a
matchup the moment its match starts.  So the Finished list — the one surface
made entirely of matches that have already started — is the one the channel
above structurally cannot serve: measured on production 2026-09-02, **118 of
its 235 rows carry no matchup at all**, only ``espn:{competition_id}``, and
every one of them rendered as dead text.

Those rows are not id-less.  ``espn:184739`` IS an id, it is the authority's,
and since lane1/057 (#2693 step 0) put an ``espn_id`` on 196 of the 200 US Open
``events`` rows there is now something for it to dereference to:

    ESPN competition id  ->  events.espn_id  ->  events.id

One indexed lookup, no name, no time window — the same posture as the market
channel, through a different id.  Measured on the same payload, 47 of the 118
resolve and the other 71 are qualifying matches for which no ``events`` row
exists at all: a coverage fact, not a join failure, and the reason the count is
published rather than the gap being papered over.

**It refuses on ambiguity, and that refusal is load-bearing.**  ``espn_id`` had
no unique constraint when this was written — 196 ids were worn by 430 rows —
so an id naming two events must resolve to neither.  Picking one would put a
reader on a coin-flip page while looking perfectly well.  The refusal is
counted as ``ESPN_ID_AMBIGUOUS``, which is also the alarm that says the step-2
repair has regressed.

Nothing here reads a probability, and nothing here writes.  Identity only —
the same rule ``tournament_register`` states about itself.
"""

from __future__ import annotations

import logging
from typing import Any, Optional, Sequence

from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession

logger = logging.getLogger(__name__)

#: Named refusals.  Every matchup that does not resolve lands in exactly one of
#: these, and the caller publishes the counts — NO SILENT CAPS (gotcha #53: a
#: zero-yield join that reports nothing reads as "there was nothing to find").
UNRESOLVED_REASONS = (
    "NO_PINNED_MARKET",
    "MARKET_NOT_FOUND",
    "MARKET_UNLINKED",
    "EVENT_DISAGREEMENT",
)

#: Named refusals of the authority-id channel.  Same rule as above: a
#: competition id that does not resolve is COUNTED, never dropped silently.
ESPN_UNRESOLVED_REASONS = (
    #: No ``events`` row carries this competition id. Almost all of these are
    #: qualifying matches we hold no market for and therefore never created.
    "NO_EVENT_FOR_ESPN_ID",
    #: Two or more rows carry it. Resolves to neither — see the module
    #: docstring, and #2693 step 2, which is what drives this to zero.
    "ESPN_ID_AMBIGUOUS",
)


def pinned_market_ids(matchup: dict[str, Any]) -> list[int]:
    """The match-winner market ids this matchup pins, across its sources.

    A ``status: "missing"`` block pins nothing by construction, so it
    contributes nothing rather than contributing a ``None`` the caller has to
    filter.  ``kind`` is checked because a register may one day pin a
    non-match market on a matchup; only the match-winner market is the thing
    whose ``event_id`` answers "which event is this".
    """
    ids: list[int] = []
    for block in matchup.get("sources") or []:
        if not isinstance(block, dict):
            continue
        if block.get("status") == "missing":
            continue
        if block.get("kind") not in (None, "match"):
            continue
        market_id = block.get("market_id")
        # `bool` is an `int` in Python and `True` would index row 1.
        if isinstance(market_id, int) and not isinstance(market_id, bool):
            ids.append(market_id)
    return ids


def _resolve_one(
    matchup: dict[str, Any], market_events: dict[int, Optional[int]]
) -> tuple[Optional[int], Optional[str]]:
    """One matchup -> ``(event_id, None)`` or ``(None, reason)``.

    Pure, so the refusal table is testable without a database.  Returns never
    both and never neither.
    """
    # A register-owned `event_id` OUTRANKS the dereference. It is a decision a
    # human wrote down against the evidence; the dereference is a decision the
    # matching layer made. When the register has spoken there is nothing to
    # resolve. (`validate_matchup` already rejects a non-positive-int here, so
    # a bad value never reaches this line.)
    pinned = matchup.get("event_id")
    if isinstance(pinned, int) and not isinstance(pinned, bool) and pinned > 0:
        return pinned, None

    market_ids = pinned_market_ids(matchup)
    if not market_ids:
        return None, "NO_PINNED_MARKET"

    found = [mid for mid in market_ids if mid in market_events]
    if not found:
        return None, "MARKET_NOT_FOUND"

    event_ids = {market_events[mid] for mid in found if market_events[mid] is not None}
    if not event_ids:
        return None, "MARKET_UNLINKED"
    if len(event_ids) > 1:
        # THE ONE CASE A NAME MATCHER WOULD GET "RIGHT" AND BE WRONG ABOUT.
        # Two pinned ids disagreeing is the register and the matching layer
        # contradicting each other about one fixture. Picking a side here would
        # bury that; refusing surfaces it as a counted finding.
        return None, "EVENT_DISAGREEMENT"

    return event_ids.pop(), None


async def resolve_matchup_events(
    session: AsyncSession, register: dict[str, Any]
) -> dict[str, Any]:
    """Every matchup in the register -> its ``events`` row, by id.

    ONE QUERY, bounded by the register.  The pinned market ids across the whole
    register are a list of at most a few hundred integers, so this is a single
    primary-key ``IN (...)`` and never a scan — the same shape as the price
    load ``get_tournament`` already runs.

    Returns::

        {
          "by_matchup":  {matchup_key: event_id},
          "by_event":    {event_id: matchup_key},
          "unresolved":  {matchup_key: reason},
          "reason_counts": {reason: n},
        }

    ``by_event`` is the direction the event page needs (it knows an id and asks
    whether it is part of a tournament) and it is built here rather than
    inverted at the call site so both surfaces read one map.  An event claimed
    by two matchups keeps neither: that is ``EVENT_DISAGREEMENT`` seen from the
    other end, and the same refusal applies.
    """
    from app.models.models import FuturesMarket  # noqa: PLC0415 — task-code import

    matchups = [m for m in (register.get("matchups") or []) if isinstance(m, dict)]

    all_ids = sorted({mid for m in matchups for mid in pinned_market_ids(m)})
    market_events: dict[int, Optional[int]] = {}
    if all_ids:
        rows = await session.execute(
            select(FuturesMarket.id, FuturesMarket.event_id).where(
                FuturesMarket.id.in_(all_ids)
            )
        )
        market_events = {int(mid): eid for mid, eid in rows.all()}

    by_matchup: dict[str, int] = {}
    unresolved: dict[str, str] = {}
    # event_id -> the matchup keys claiming it, so a double claim is visible
    # rather than last-write-wins.
    claims: dict[int, list[str]] = {}

    for matchup in matchups:
        key = matchup.get("matchup_key")
        if not isinstance(key, str) or not key:
            continue
        event_id, reason = _resolve_one(matchup, market_events)
        if event_id is None:
            unresolved[key] = reason or "NO_PINNED_MARKET"
            continue
        by_matchup[key] = event_id
        claims.setdefault(event_id, []).append(key)

    by_event: dict[int, str] = {}
    for event_id, keys in claims.items():
        if len(keys) == 1:
            by_event[event_id] = keys[0]
            continue
        # One event, two matchups: the same event offered as two different
        # matches. Drop every claimant — see the module docstring.
        for key in keys:
            by_matchup.pop(key, None)
            unresolved[key] = "EVENT_DISAGREEMENT"
        logger.warning(
            "tournament event link: event %s claimed by %s matchups (%s) — all dropped",
            event_id, len(keys), ", ".join(sorted(keys)),
        )

    reason_counts: dict[str, int] = {}
    for reason in unresolved.values():
        reason_counts[reason] = reason_counts.get(reason, 0) + 1

    return {
        "by_matchup": by_matchup,
        "by_event": by_event,
        "unresolved": unresolved,
        "reason_counts": dict(sorted(reason_counts.items())),
    }


async def resolve_espn_competition_events(
    session: AsyncSession,
    competition_ids: Sequence[str],
    sport_keys: Sequence[str],
) -> dict[str, Any]:
    """ESPN competition id -> ``events.id``, for the ids a results feed names.

    ONE QUERY, bounded twice over: by the competition ids the caller actually
    holds (a few hundred at most) and by ``sport_keys``, which is the
    tournament spec's own named list.  The sport bound is not an optimisation —
    it is what stops a tennis competition id from resolving to a baseball row
    should the two id spaces ever collide, and it is named in
    ``REGISTERED_TOURNAMENTS`` rather than inferred from the slug for the same
    reason every other bound on this page is.

    Returns::

        {
          "by_espn":       {competition_id: event_id},
          "unresolved":    {competition_id: reason},
          "reason_counts": {reason: n},
        }

    An id carried by two events resolves to NEITHER.  ``espn_id`` is not unique
    yet (#2693 step 2), and a link that guesses between two rows is worse than
    no link: it is wrong half the time and looks right every time.
    """
    from app.models.models import Event, Sport  # noqa: PLC0415 — task-code import

    wanted = [str(cid) for cid in dict.fromkeys(competition_ids) if cid]
    by_espn: dict[str, int] = {}
    unresolved: dict[str, str] = {}
    if not wanted or not sport_keys:
        return {"by_espn": by_espn, "unresolved": unresolved, "reason_counts": {}}

    rows = await session.execute(
        select(Event.espn_id, Event.id)
        .join(Sport, Sport.id == Event.sport_id)
        .where(Event.espn_id.in_(wanted), Sport.key.in_(list(sport_keys)))
    )

    claims: dict[str, list[int]] = {}
    for espn_id, event_id in rows.all():
        claims.setdefault(str(espn_id), []).append(int(event_id))

    for competition_id in wanted:
        found = claims.get(competition_id) or []
        if not found:
            unresolved[competition_id] = "NO_EVENT_FOR_ESPN_ID"
        elif len(set(found)) > 1:
            unresolved[competition_id] = "ESPN_ID_AMBIGUOUS"
            logger.warning(
                "tournament event link: espn competition %s names %s events (%s) — "
                "all dropped; #2693 step 2 has regressed",
                competition_id, len(set(found)), ", ".join(str(e) for e in sorted(set(found))),
            )
        else:
            by_espn[competition_id] = found[0]

    reason_counts: dict[str, int] = {}
    for reason in unresolved.values():
        reason_counts[reason] = reason_counts.get(reason, 0) + 1

    return {
        "by_espn": by_espn,
        "unresolved": unresolved,
        "reason_counts": dict(sorted(reason_counts.items())),
    }

File: frontend/lib/tournamentEventLink.ts
/**
 * WHICH EVENT PAGE DOES THIS MATCHUP ROUTE TO — one answer, read by every list
 * on the tournament hub (ux/1002).
 *
 * ═══ WHY THIS IS ITS OWN MODULE ═══
 *
 * Alex, on the live hub at 10pm 2026-09-01: *"the Zverev–Sonego card (LIVE,
 * '4TH SET') is not clickable. CERT-703 wired only the FINISHED list. The API
 * already provides the mapping (event_links.by_matchup -> 15293811)."*
 *
 * The diagnosis names the real structural fault, which outlives whichever card
 * happened to be dead when it was read: **the hub had two different ideas of
 * where a match links to.** The FINISHED list read the payload's published map,
 * `event_links.by_matchup` — one id-anchored answer per matchup, resolved
 * server-side by `utils/tournament_event_link.py`. The MATCH list read
 * something else entirely: a per-row `event_id` stamped onto slate rows only.
 *
 * Those two sources agree on the rows they both cover, which is what makes the
 * split so easy to miss and so bad to keep. They stop agreeing exactly where it
 * costs a reader a click:
 *
 *   - **A name-pair join is the join this product does not do.** Gotcha #32 /
 *     ruling 048 is written against exactly this: identity travels by id. The
 *     matchup key is the id both halves of the payload already agree on.
 *
 * So the resolution moves here, both lists call it, and the rule about where a
 * match links has one definition instead of two implementations.
 *
 * ═══ WHAT THIS MODULE DOES **NOT** BUY, MEASURED (ux/1008, CERT-724) ═══
 *
 * Round one also claimed it links cards that were dead. It does not, and the
 * correction belongs here rather than in a report, because the directive that
 * produced the claim is still on file and reads persuasively:
 *
 *   - **For a SLATE row the fallback is unreachable.** `build_slate` already
 *     fills `event_id` from this very map when the register pins nothing
 *     (`tournament_slate.py:692`), so "in `by_matchup`" implies "stamped on the
 *     row". Measured on the captured payload, rendered through the real
 *     component: map and no-map produce the identical set of ten hrefs, and the
 *     live card Alex reported as dead is an anchor under BOTH rules.
 *   - **For a BRACKET row it is unreachable too, and for a different reason.**
 *     `matchListFromBracket` nulls `matchupKey` and `eventId` together, so the
 *     row has no key to look up. Left unfixed deliberately:
 *     `ingest_espn_draw.py` never writes `draw_slot`, `build_bracket` returns
 *     `[]`, and production ships no bracket rows at all.
 *
 * The module earns its place as the ONE definition of the rule, not as a
 * source of new links. The `espn:` refusal below is the part with real teeth.
 *
 * ═══ WHAT IT REFUSES, AND WHY EACH REFUSAL IS LOAD-BEARING ═══
 *
 * A missing link is a smaller harm than a wrong one — the server module says so
 * about its own four refusals, and the client must not be the layer that
 * relaxes them:
 *
 *   - **No key in the map.** The server resolved nothing for this matchup
 *     (`NO_PINNED_MARKET`, `MARKET_UNLINKED`, `MARKET_NOT_FOUND`,
 *     `EVENT_DISAGREEMENT`) and published the reason counts beside the map. We
 *     do not get to guess past a refusal that was counted.
 *   - **An `espn:` key NEVER resolves.** This is not a formatting quirk, it is
 *     Q503/Q505's whole point. When the register's pairing disagrees with the
 *     scoreboard, `authority_match_row` rebuilds the card with the authority's
 *     two players and no price, and re-keys it `espn:{competition_id}`
 *     precisely so it cannot reach a consumer that keys on the register's
 *     matchup — whose event page would print the pairing we just withheld. The
 *     register key for that same fixture is very much in `by_matchup`; looking
 *     it up would hand the reader a link to the lie. The prefix test is the
 *     guard, and it is why this function takes the KEY and not the row.
 *   - **A non-positive or non-finite id.** JSON round-trips through Redis and
 *     an id is only an id if it could address a row.
 */

/**
 * ═══ THE SECOND MAP, AND WHY IT IS NOT A RELAXATION OF THE FIRST (#2693) ═══
 *
 * The `espn:` refusal above stays exactly as written. What follows is a
 * different question asked of a different map, and the distinction is the
 * whole of its safety.
 *
 * `matchupEventId` refuses an `espn:` key because the destination it would
 * reach is *the register's* event for that matchup — the one holding the
 * pairing Q503 just withheld. `espnCompetitionEventId` does not go there. It
 * reads `event_links.by_espn`, a map the server builds by dereferencing the
 * AUTHORITY's competition id through `events.espn_id`, so the row it lands on
 * is one that agrees with the authority by construction: lane1/057's anchor
 * join only stamps an `espn_id` on an event whose two players ESPN confirms.
 *
 * So the row Q503 re-keyed to `espn:184739` now links — to ESPN's match, not to
 * the register's. The refusal above was never "this row may not be linked"; it
 * was "this row may not be linked THROUGH THE REGISTER'S KEY", and it still is.
 *
 * Order is market-first, authority-second. The market channel is the reviewed
 * one, and in the disagreement case it returns `null` anyway (the row's key is
 * `espn:`), so the fallback can never overrule a link a human pinned.
 */

/** The published map: register matchup key -> our `events.id`. */
export type MatchupEventIds = Record<string, number> | null | undefined;

/** The published map: ESPN competition id -> our `events.id`. */
export type EspnEventIds = Record<string, number> | null | undefined;

/** An id is only an id if it could address a row. Shared by both channels. */
function usableEventId(value: unknown): number | null {
  if (typeof value !== "number" || !Number.isFinite(value) || value <= 0) {
    return null;
  }
  return value;
}

/**
 * The `events.id` for a matchup key, or `null` — the ONE resolution.
 *
 * Pure and total: every refusal above returns `null` rather than throwing, so a
 * malformed payload costs a link and never a render.
 */
export function matchupEventId(
  matchupKey: string | null | undefined,
  eventIds: MatchupEventIds
): number | null {
  if (!eventIds) return null;
  if (typeof matchupKey !== "string" || matchupKey.length === 0) return null;
  // The authority-named row. See the docstring — this is the refusal that
  // stops a correct-looking link from being a wrong one.
  if (matchupKey.startsWith("espn:")) return null;
  return usableEventId(eventIds[matchupKey]);
}

/** `/events/{id}`, or `null`. The standard event page; never a hub-private URL. */
export function matchupEventHref(
  matchupKey: string | null | undefined,
  eventIds: MatchupEventIds
): string | null {
  const eventId = matchupEventId(matchupKey, eventIds);
  return eventId === null ? null : `/events/${eventId}`;
}

/**
 * The `events.id` for an ESPN competition id, or `null` — the second channel.
 *
 * Pure and total, like its sibling. A competition id the server could not
 * resolve is simply absent from the map: it published
 * `NO_EVENT_FOR_ESPN_ID` (no row exists — most of the qualifying draw) or
 * `ESPN_ID_AMBIGUOUS` (two rows carry it, so neither is the answer) beside it.
 * The client does not get to guess past a refusal that was counted, on this
 * channel any more than on the other.
 */
export function espnCompetitionEventId(
  espnCompetitionId: string | number | null | undefined,
  espnEventIds: EspnEventIds
): number | null {
  if (!espnEventIds) return null;
  if (espnCompetitionId === null || espnCompetitionId === undefined) return null;
  const key = String(espnCompetitionId);
  if (key.length === 0) return null;
  return usableEventId(espnEventIds[key]);
}

/**
 * Where a match row links — BOTH channels, in the one order that is safe.
 *
 * Market first, authority second. See the block comment above `MatchupEventIds`
 * for why the fallback cannot resurrect the link Q503 withheld.
 */
export function matchEventHref(
  matchupKey: string | null | undefined,
  espnCompetitionId: string | number | null | undefined,
  eventIds: MatchupEventIds,
  espnEventIds?: EspnEventIds
): string | null {
  const pinned = matchupEventId(matchupKey, eventIds);
  if (pinned !== null) return `/events/${pinned}`;
  const viaAuthority = espnCompetitionEventId(espnCompetitionId, espnEventIds);
  return viaAuthority === null ? null : `/events/${viaAuthority}`;
}
